### PR TITLE
libav: mark as insecure

### LINKED ISSUE #90839: Vulnerability roundup 85: libav-0.8.21: 16 advisories [9.8]
[search](https://search.nix.gsc.io/?q=libav&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=libav+in%3Apath&type=Code)

* [ ] [CVE-2017-9051](https://nvd.nist.gov/vuln/detail/CVE-2017-9051) CVSSv3=9.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2016-3062](https://nvd.nist.gov/vuln/detail/CVE-2016-3062) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-5684](https://nvd.nist.gov/vuln/detail/CVE-2018-5684) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-5766](https://nvd.nist.gov/vuln/detail/CVE-2018-5766) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9719](https://nvd.nist.gov/vuln/detail/CVE-2019-9719) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2017-16803](https://nvd.nist.gov/vuln/detail/CVE-2017-16803) CVSSv3=7.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2015-5479](https://nvd.nist.gov/vuln/detail/CVE-2015-5479) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9717](https://nvd.nist.gov/vuln/detail/CVE-2019-9717) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9720](https://nvd.nist.gov/vuln/detail/CVE-2019-9720) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2016-6832](https://nvd.nist.gov/vuln/detail/CVE-2016-6832) CVSSv3=5.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2016-7393](https://nvd.nist.gov/vuln/detail/CVE-2016-7393) CVSSv3=5.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2016-7424](https://nvd.nist.gov/vuln/detail/CVE-2016-7424) CVSSv3=5.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2016-8675](https://nvd.nist.gov/vuln/detail/CVE-2016-8675) CVSSv3=5.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2016-8676](https://nvd.nist.gov/vuln/detail/CVE-2016-8676) CVSSv3=5.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2014-5271](https://nvd.nist.gov/vuln/detail/CVE-2014-5271) (nixos-20.03, nixos-unstable)
* [ ] [CVE-2015-3395](https://nvd.nist.gov/vuln/detail/CVE-2015-3395) (nixos-20.03, nixos-unstable)

Scanned versions: nixos-20.03: a84b797b28e; nixos-unstable: 22c98819ccd. May contain false positives.

### LINKED ISSUE #90844: Vulnerability roundup 85: libav-11.12: 6 advisories [9.8]
[search](https://search.nix.gsc.io/?q=libav&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=libav+in%3Apath&type=Code)

* [ ] [CVE-2017-9051](https://nvd.nist.gov/vuln/detail/CVE-2017-9051) CVSSv3=9.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-5684](https://nvd.nist.gov/vuln/detail/CVE-2018-5684) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-5766](https://nvd.nist.gov/vuln/detail/CVE-2018-5766) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9719](https://nvd.nist.gov/vuln/detail/CVE-2019-9719) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9717](https://nvd.nist.gov/vuln/detail/CVE-2019-9717) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9720](https://nvd.nist.gov/vuln/detail/CVE-2019-9720) CVSSv3=6.5 (nixos-20.03, nixos-unstable)

Scanned versions: nixos-20.03: a84b797b28e; nixos-unstable: 22c98819ccd. May contain false positives.

### LINKED ISSUE #90852: Vulnerability roundup 85: libav-12.3: 18 advisories [8.8]
[search](https://search.nix.gsc.io/?q=libav&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=libav+in%3Apath&type=Code)

* [ ] [CVE-2019-9719](https://nvd.nist.gov/vuln/detail/CVE-2019-9719) CVSSv3=8.8 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-11102](https://nvd.nist.gov/vuln/detail/CVE-2018-11102) CVSSv3=7.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-11224](https://nvd.nist.gov/vuln/detail/CVE-2018-11224) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-18826](https://nvd.nist.gov/vuln/detail/CVE-2018-18826) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-18827](https://nvd.nist.gov/vuln/detail/CVE-2018-18827) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-18828](https://nvd.nist.gov/vuln/detail/CVE-2018-18828) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-18829](https://nvd.nist.gov/vuln/detail/CVE-2018-18829) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-19128](https://nvd.nist.gov/vuln/detail/CVE-2018-19128) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-19129](https://nvd.nist.gov/vuln/detail/CVE-2018-19129) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-19130](https://nvd.nist.gov/vuln/detail/CVE-2018-19130) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2018-20001](https://nvd.nist.gov/vuln/detail/CVE-2018-20001) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9717](https://nvd.nist.gov/vuln/detail/CVE-2019-9717) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-9720](https://nvd.nist.gov/vuln/detail/CVE-2019-9720) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-14371](https://nvd.nist.gov/vuln/detail/CVE-2019-14371) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-14372](https://nvd.nist.gov/vuln/detail/CVE-2019-14372) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-14441](https://nvd.nist.gov/vuln/detail/CVE-2019-14441) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-14442](https://nvd.nist.gov/vuln/detail/CVE-2019-14442) CVSSv3=6.5 (nixos-20.03, nixos-unstable)
* [ ] [CVE-2019-14443](https://nvd.nist.gov/vuln/detail/CVE-2019-14443) CVSSv3=6.5 (nixos-20.03, nixos-unstable)

Scanned versions: nixos-20.03: a84b797b28e; nixos-unstable: 22c98819ccd. May contain false positives.

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
closes https://github.com/NixOS/nixpkgs/issues/90844 and closes https://github.com/NixOS/nixpkgs/issues/90852 and closes https://github.com/NixOS/nixpkgs/issues/90839
I didn't list all vulnerabilities under `knownVulnerabilities`.

###### Consequences of this change
This disables the following packages:
```
7 packages removed:
clickshare-csc1 (†01.07.00.033) keyfinder (†2.2) libav (†0.8.21) libav (†11.12) libav (†12.3) vdr-markad (†2017-03-13) vdr-xineliboutput (†2.2.0)
```
~Is there a way to make ofborg ping their maintainers?~

For those maintainers: Consider switching to ffmpeg.
- [x] for beets cc @doronbehar @lovesegfault @austinbutler @mweinelt (fixed by updating keyfinder-cli, please test)
- [ ] for clickshare-csc1 cc @Yarny0 
- [x] for gerbera cc @ardumont
- [x] for guitarix cc @orivej @magnetophon~ (I don't see any dependency on libav/ffmpeg in other distros and the output of the configurePhase)
- [x] for jabref cc @wamserma (doesn't actually depend on libav/ffmpeg)
- [x] for musly cc @ggPeti (fixed by replacing libav with ffmpeg, please test)
- [x] for oraclejdk @lourkeur @Ma27 (fixed in #111528)
- [x] for signumone-ks cc @WolfangAukang 
- [x] for steam cc @Atemu @tadfisher @abbradar @jonringer @danieldk @bbigras @Flakebi @nyanloutre
- [x] for untrunc cc @erikarvstedt @peti (fixed by making it an alias of untrunc-anthwlock)
- [ ] for vdr-markad and vdr-xineliboutput cc @ck3d
- [x] for zulu cc @fpletz @yaroot

### CODE REVIEW

**Comment by erictapen** [comment]
Thats a lot of packages to break. Why don't we patch for the vulnerabilities?

**Comment by dotlambda** [comment]
> Thats a lot of packages to break. Why don't we patch for the vulnerabilities?

I guess we could find some patches over at ffmpeg, but wouldn't it be easier to patch everything to use ffmpeg instead?

**Comment by vcunat** [comment]
I'm not aware of a way to notify them.  `steam*` is certainly used; I'm not sure how important libav is in there and perhaps ffmpeg is perfectly OK.

**Comment by dotlambda** [comment]
@jtojnar @IvarWithoutBones @prusnak Libav is not in https://github.com/AppImage/pkg2appimage/blob/master/excludelist. Should we just remove it from `appimageTools`'s dependencies or replace it by ffmpeg?

**Comment by dotlambda** [comment]
> Thats a lot of packages to break.

I brought the number down from 74 to 30.

**Comment by rycee** [comment]
Checked the unpaper change and it is OK. Thanks!

**Comment by twhitehead** [comment]
My only experience is with the ovito package. I changed libav in the `buildInputs` to  ffmpeg and it  built fine, so certainly seems worthwhile trying the same with any of these other packages that people may be using.

File: pkgs/development/libraries/libav/default.nix [hunk 1]
```diff
@@ -127,6 +127,10 @@ let
       license = with licenses; if enableUnfree then unfree #ToDo: redistributable or not?
         else if enableGPL then gpl2Plus else lgpl21Plus;
       platforms = with platforms; linux ++ darwin;
+      knownVulnerabilities =
+        lib.optional (lib.versionOlder version "12.1") "CVE-2017-9051"
+        ++ lib.optionals (lib.versionOlder version "12.3") [ "CVE-2018-5684" "CVE-2018-5766" ]
+        ++ lib.optionals (lib.versionOlder version "12.4") [ "CVE-2019-9717" "CVE-2019-9720" ];
```
Comment (twhitehead) [inline pkgs/development/libraries/libav/default.nix:133]: The complete list from the links you provided is

CVE-2014-5271
CVE-2015-3395
CVE-2015-5479
CVE-2016-3062
CVE-2016-6832
CVE-2016-7393
CVE-2016-7424
CVE-2016-8675
CVE-2016-8676
CVE-2017-16803
CVE-2017-9051
CVE-2018-11102
CVE-2018-11224
CVE-2018-18826
CVE-2018-18827
CVE-2018-18828
CVE-2018-18829
CVE-2018-19128
CVE-2018-19129
CVE-2018-19130
CVE-2018-20001
CVE-2018-5684
CVE-2018-5766
CVE-2019-14371
CVE-2019-14372
CVE-2019-14441
CVE-2019-14442
CVE-2019-14443
CVE-2019-9717
CVE-2019-9719
CVE-2019-9720

Definitely makes the case that this package should be gone. I guess the thinking then is it isn't really worth the time to lay these all out? I'm okay with that.

**Review by twhitehead — APPROVED** [review]
That's a huge list of CVEs with no hope of being fixed upstream. With ffmpeg being pretty much a drop in replacement,  makes sense to start the process of phasing out its usage.

**Comment by dotlambda** [comment]
Alternative suggestion: Just use `libav` as an alias for `ffmpeg` and drop the expressions for libav. This will cause some packages not to build which should also catch the maintainers' attention.

**Comment by dotlambda** [comment]
@erikarvstedt @peti @romildo  Can we make `untrunc` an alias of `untrunc-anthwlock`?

**Comment by romildo** [comment]
> @erikarvstedt @peti @romildo Can we make `untrunc` an alias of `untrunc-anthwlock`?

It is ok for me. [`untrunc-anthwlock`](https://github.com/anthwlock/untrunc) has more features than [`untrunc`](https://github.com/ponchio/untrunc).

**Comment by dotlambda** [comment]
I tried running `keyfinder` and only get
```
$ result/bin/KeyFinder
[1]    2610 abort (core dumped)  ./result/bin/KeyFinder
```
So I suggest removing it completely.

**Comment by lourkeur** [comment]
#111528 merged, oraclejdk8 and referrers can be ticked off.

**Comment by wamserma** [comment]
@dotlambda droppping libav from the jabref expression LGTM (tested locally).

**Comment by vcunat** [comment]
> 2021-02-01 15:59:04.016 Fatal: This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

I think (some) Qt stuff was throwing this unless you installed it (via `nix-env` or `systemPackages`).

**Comment by dotlambda** [comment]
> I think (some) Qt stuff was throwing this unless you installed it (via `nix-env` or `systemPackages`).

I tried `nix-env -iA keyfinder` as well.

**Comment by twhitehead** [comment]
@dotlambda I think that Qt error typically happens when the QT_PLUGIN_PATH hasn't been set so Qt can find the appropriate backend module (i.e., xcb in the X11 backend).

Typically this is caused by using `stdenv.mkDerivation` instead of `mkDerivation`. The later is brought into scope with the `libsForQtX.callPackage` call and includes an automatic wrapping system. See [the Qt section of the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#qt-default-nix-co-2) for details.

**Comment by dotlambda** [comment]
> @dotlambda I think that Qt error typically happens when the QT_PLUGIN_PATH hasn't been set so Qt can find the appropriate backend module (i.e., xcb in the X11 backend).
> 
> Typically this is caused by using `stdenv.mkDerivation` instead of `mkDerivation`. The later is brought into scope with the `libsForQtX.callPackage` call and includes an automatic wrapping system. See [the Qt section of the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#qt-default-nix-co-2) for details.

I'm not saying that `keyfinder` can't be fixed. But since it's been probably been broken for quite a while I see no harm in removing it from nixpkgs.

**Comment by dotlambda** [comment]
It would be nice if someone using steam tried replacing `libav` with `ffmpeg` and reported back whether Steam still works fine.
cc @Atemu @tadfisher @abbradar @jonringer @danieldk @bbigras @Flakebi @nyanloutre

**Comment by jonringer** [comment]
i can do it

**Comment by jonringer** [comment]
I was able to launch csgo and factorio fine with the following diff:
```
diff --git a/pkgs/games/steam/fhsenv.nix b/pkgs/games/steam/fhsenv.nix
index 4fa8960726b..b3ed711805e 100644
--- a/pkgs/games/steam/fhsenv.nix
+++ b/pkgs/games/steam/fhsenv.nix
@@ -171,7 +171,7 @@ in buildFHSUserEnv rec {
     SDL2
     libusb1
     dbus-glib
-    libav
+    ffmpeg
     atk
     # Only libraries are needed from those two
     libudev0-shim
     ```

**Comment by jonringer** [comment]
however, the weird thing is that it's not changing the build ... but it appears that libav doesn't appear in the runtime closure anyway
```
$ nix-store --query --tree $(nix-build '<nixpkgs>' -A steam) | grep libav
[09:10:14] jon@jon-desktop /home/jon/projects/nixpkgs (master)
```

**Comment by jonringer** [comment]
oh, I know why, those only get brought in if you have the "native runtime". Otherwise it will just give you a nixified steam-scout runtime

**Comment by twhitehead** [comment]
@dotlambda using `mkDerivation` instead of `stdenv.mkDerivation` makes `keyfinder` run.

Unfortunately, even though I tried a variety of things, such as all versions of `ffmpeg`, and some minimal patching to fix the initial errors, I could not get it to compile against `ffmpeg` without having to invest more time than I care to.

There is an [open issue about it upstream](https://github.com/ibsh/is_KeyFinder/issues/110), so I added a further comment there.  I say this one gets let go as you suggested.

**Comment by jonringer** [comment]
I was able to launch csgo with `steam-run-native` and the patch from above applied.

**Review by jonringer — APPROVED** [review]

**Comment by magnetophon** [comment]
@dotlambda Thanks for the fix!

**Review by ggPeti — APPROVED** [review]

**Review by wamserma — APPROVED** [review]

**Comment by vcunat** [comment]
Steam: I'm not really familiar with how their runtime works, but if I look at what it currently installs:
```
$ (cd ~/.local/share/Steam/ && find -name '*libav*')
./ubuntu12_32/libavcodec.so.57
./ubuntu12_32/libavformat.so.57
./ubuntu12_32/libavresample.so.3
./ubuntu12_32/libavutil.so.55
./steamapps/common/**
```
the versions look like `ffmpeg_3` and not `ffmpeg_4`:
```
$ (cd $(nix-build -QA ffmpeg_3.out)/lib && ls *.so.{?,??}) | cat
libavcodec.so.57
libavdevice.so.57
libavfilter.so.6
libavformat.so.57
libavresample.so.3
libavutil.so.55
libpostproc.so.54
libswresample.so.2
libswscale.so.4
```

**Comment by dotlambda** [comment]
There are only four (families of) packages left that depend on some version of Libav:
- [ ] clickshare-csc1
- [ ] signumone-ks
- [ ] vdr-markad and vdr-xineliboutput
- [ ] zulu

**Comment by vcunat** [comment]
`meta.knownVulnerabilities` breaks evaluation transitively, so I expect it's OK to leave some cases as they are.

**Comment by dotlambda** [comment]
> it's OK to leave some cases as they are

Then I think this is ready to be merged.

**Comment by WolfangAukang** [comment]
#111693 for signumone-ks. Removing `libav_0_8`

**Comment by yaroot** [comment]
@dotlambda I don't think zulu should depend on libav or ffmpeg, I've checked all the so files it should be fine to remove them.

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -294,6 +294,7 @@ mapAliases ({
   kinetic-cpp-client = throw "kinetic-cpp-client has been removed from nixpkgs, as it's abandoned."; # 2020-04-28
   kicad-with-packages3d = kicad; # added 2019-11-25
   krename-qt5 = krename; # added 2017-02-18
+  keyfinder = throw "keyfinder has been removed because it was broken"; # added 2021-02-01
```
Comment (ajs124) [inline pkgs/top-level/aliases.nix:297]: It seems to work for someone https://github.com/NixOS/nixpkgs/issues/18475#issuecomment-771504453

Comment (dotlambda) [reply pkgs/top-level/aliases.nix:297]: False positive: https://github.com/NixOS/nixpkgs/issues/18475#issuecomment-772613652 ;-)

Comment (dotlambda) [reply pkgs/top-level/aliases.nix:297]: @ajs124 Or would say we should keep it anyway? It would defiitely be marked as insecure.

Comment (dotlambda) [reply pkgs/top-level/aliases.nix]: Instead of removing it, I now added a commit updating it to version 2.4.

**Comment by Yarny0** [comment]
Thanks for the notification about clickshare-csc1.  Sadly, clickshare won't build with ffmpeg: It needs `libswscale.so.2` which is not provided by any of the ffmpeg versions currently in nixpkgs.  I suspect it would need a much older ffmpeg version which we no longer package.  Sadly, there is no hope for an update of clickshare as the company no longer produces a linux version.

As long as libav_0_8 isn't removed from nixpkgs, it's probably best to leave clickshare as it is.  Users are warned about the situation and can decide themselves if they accept the risk of installing a vulnerable package.  I don't know of any other solution.

**Comment by dotlambda** [comment]
Is anything preventing us from merging?

**Review by ggPeti — APPROVED** [review]

**Comment by dotlambda** [comment]
backport: https://github.com/NixOS/nixpkgs/pull/112396

**Pull request merged**